### PR TITLE
[CRB-262] Implement reward block list request

### PR DESCRIPTION
**Pull request by nathanwhit**
The handler itself is essentially copy-pasted from CC 1.7 (that part of sawtooth didn't change much between 1.0.5 and 1.2). The rest of the PR threads through the `block_signature` to pass to the TP during validation. This was not yet ported from CC 1.7, which is why we haven't run into problems in testing up to this point.

### CODE REVIEW

**Review by seandeneen-gluwa — APPROVED** [review]

File: validator/sawtooth_validator/state/client_handlers.py [hunk 1]
```diff
@@ -778,6 +779,116 @@ def _respond(self, request):
...
+
+        height = head_block.block_num
+        is_block = self.BoolWrapper()
+
+        try:
+            #find tip to be rewarded
+            while head_block.block_num != first_pred_height and head_block.block_num == height:
+                if head_block.header_signature in self._block_store:
+                    is_block.found_in_store = True
+                head_block = BlockWrapper(next(self._block_manager.get([head_block.previous_block_id])))
```
Comment (CAGS295) [inline validator/sawtooth_validator/state/client_handlers.py]:
```suggestion
                head_block = BlockWrapper(next(self._block_manager.get([head_block.previous_block_id])))
                if head_block.header_signature in self._block_store:
                    found_in_store = True
```

File: validator/sawtooth_validator/state/client_handlers.py [hunk 2]
```diff
@@ -778,6 +779,116 @@ def _respond(self, request):
...
+        super().__init__(
+            client_block_pb2.ClientRewardBlockListRequest,
+            client_block_pb2.ClientRewardBlockListResponse,
+            validator_pb2.Message.CLIENT_REWARD_BLOCK_LIST_RESPONSE,
+            block_store = block_store)
+        self._block_manager = block_manager
+
+    class BoolWrapper:
+        def __init__(self) -> None:
+            self.found_in_store=False
```
Comment (CAGS295) [inline validator/sawtooth_validator/state/client_handlers.py]:
```suggestion
```

File: validator/sawtooth_validator/state/client_handlers.py [hunk 3]
```diff
@@ -778,6 +779,116 @@ def _respond(self, request):
...
+
+        try:
+            #find tip to be rewarded
+            while head_block.block_num != first_pred_height and head_block.block_num == height:
+                if head_block.header_signature in self._block_store:
+                    is_block.found_in_store = True
+                head_block = BlockWrapper(next(self._block_manager.get([head_block.previous_block_id])))
+                height -= 1
+                #predecessors are in the store already
+                if is_block.found_in_store:
```
Comment (CAGS295) [inline validator/sawtooth_validator/state/client_handlers.py]:
```suggestion
                if found_in_store:
```

File: validator/sawtooth_validator/state/client_handlers.py (hunk 2 above)
Comment (CAGS295) [reply validator/sawtooth_validator/state/client_handlers.py]: Not needed anymore.

File: validator/sawtooth_validator/state/client_handlers.py (hunk 1 above)
Comment (CAGS295) [reply validator/sawtooth_validator/state/client_handlers.py]: It is a bit confusing that if the branch holds true, the next block query has no effect. Consider querying first and then checking if the block is in the store.

File: validator/sawtooth_validator/state/client_handlers.py [hunk 4]
```diff
@@ -778,6 +779,116 @@ def _respond(self, request):
...
+
+        first_pred_height = request.first_predecessor_height
+        last_pred_height = request.last_predecessor_height
+
+        padding_traversal = True
+        # Fetch the foreign head block.
+        try:
+            head_block = BlockWrapper(next(self._block_manager.get([request.head_id])))
+        except StopIteration:
+            LOGGER.debug('Unable to find block "%s" in cache', request.head_id)
```
Comment (CAGS295) [inline validator/sawtooth_validator/state/client_handlers.py]:
```suggestion
            LOGGER.debug('Unable to find block "%s" in the block manager', request.head_id)
```

File: validator/sawtooth_validator/state/client_handlers.py [hunk 5]
```diff
@@ -778,6 +779,116 @@ def _respond(self, request):
...
+
+        if not padding_traversal:
+            try:
+                head_block = self._block_store.get_block_by_number(first_pred_height)
+            except Exception as e:
+                LOGGER.debug("Unable to fetch first predecessor block in store by blocknum")
+                raise _ResponseFailed(self._status.INTERNAL_ERROR)
+
+        height = head_block.block_num
+        is_block = self.BoolWrapper()
```
Comment (CAGS295) [inline validator/sawtooth_validator/state/client_handlers.py]:
```suggestion
        found_in_store = False
```

**Comment by atodorov** [comment]
@nathanwhit this PR has some changes requested. Ideally we should retest with the latest changes before merging.

**Comment by nathanwhit** [comment]
Addressed comments, ready for retesting @atodorov

**Comment by atodorov** [comment]
@CAGS295 can you review again before I go ahead and redeploy everything?

**Review by atodorov — APPROVED** [review]
Note to self: not related to the current issues we're seeing. Can merge.

**Review by CAGS295 — APPROVED** [review]
LGTM

**Pull request merged**